src/Storages/PostgreSQL/PostgreSQLConnectionPool.cpp

```diff
@@ -0,0 +1,77 @@
+#if !defined(ARCADIA_BUILD)
+#include "config_core.h"
+#endif
+
+#if USE_LIBPQXX
+#include <IO/WriteBufferFromString.h>
+#include <IO/Operators.h>
+#include "PostgreSQLConnectionPool.h"
+#include "PostgreSQLConnection.h"
+#include <common/logger_useful.h>
+
+
+namespace DB
+{
+
+PostgreSQLConnectionPool::PostgreSQLConnectionPool(
+    std::string dbname, std::string host, UInt16 port, std::string user, std::string password)
+{
+    address = host + ':' + std::to_string(port);
+    connection_str = formatConnectionString(std::move(dbname), std::move(host), port, std::move(user), std::move(password));
+    initialize();
+}
+
+
+PostgreSQLConnectionPool::PostgreSQLConnectionPool(const PostgreSQLConnectionPool & other)
+        : connection_str(other.connection_str)
+        , address(other.address)
+{
+    initialize();
+}
+
+
+std::string PostgreSQLConnectionPool::formatConnectionString(
+    std::string dbname, std::string host, UInt16 port, std::string user, std::string password)
+{
+    WriteBufferFromOwnString out;
+    out << "dbname=" << quote << dbname
+        << " host=" << quote << host
+        << " port=" << port
+        << " user=" << quote << user
+        << " password=" << quote << password;
+    return out.str();
+}
+
+
+void PostgreSQLConnectionPool::initialize()
+{
+    /// No connection is made, just fill pool with non-connected connection objects.
+    for (size_t i = 0; i < POSTGRESQL_POOL_DEFAULT_SIZE; ++i)
+        pool.emplace_back(std::make_shared<PostgreSQLConnection>(connection_str, address));
+}
+
+
+WrappedPostgreSQLConnectionPtr PostgreSQLConnectionPool::get()
+{
+    std::lock_guard lock(mutex);
+
+    for (const auto & connection : pool)
+    {
+        if (connection->isAvailable())
```

Comment:
So basically each connection has a 'locked' flag that is set when you give it away, and unset when the client doesn't need it anymore. And you're iterating here over all the connections to find one that is not used.

Looks like there is an easier way. When you give a connection away, remove it from the `pool` array. Return a wrapper that will put the connection back into the pool in destructor. Better name it something like `PostgreSQLPooledConnectionHolder`, it is still very generic but better than `Wrapped`. This way, you won't need any additional code in the `PostgreSQLConnection`, and won't have any shared access to these connections.